velox/expression/ComplexProxyTypes.h

```diff
@@ -228,4 +233,166 @@ class ArrayProxy {
   template <typename T>
   friend class SimpleFunctionAdapter;
 };
+
+// The object passed to the simple function interface that represent a single
+// output map entry.
+//
+// General Interface:
+// - add_item()  : return references to key and value writers as tuple.
+// - add_null()  : return key writer, set value to null.
+// - size()      : return the size of the map.
+// The interface does not guarantee that duplicates not written.
+template <typename K, typename V>
+class MapWriter {
+  using key_writer_t = VectorWriter<K, void>;
+  using value_writer_t = VectorWriter<V, void>;
+
+  using key_element_t = typename key_writer_t::exec_out_t;
+  using value_element_t = typename value_writer_t::exec_out_t;
+
+  // Provide std::interface only if both key and value are eligible types.
+  static constexpr bool std_interface =
+      provide_std_interface<K> && provide_std_interface<V>;
+
+ public:
+  // `size` is with respect to the current size of the map being written.
+  void reserve(vector_size_t size) {
+    auto currentSize = size + innerOffset_;
+
+    if (UNLIKELY(currentSize > capacity_)) {
+      while (currentSize > capacity_) {
+        capacity_ = capacity_ * 2;
+      }
```

Comment:
why not do it without looping?

capacity_ *= std::pow(2, std::ceil(std::log(currentSize/capacity_)))

I don't know what the performance of that would be, but if currentSize is >> capacity_ I imagine it would be better

Reply:
capacity_ = std::pow(2, std::ceil(std::log2(currentSize)));
      worked and sounds better.